ocaml/xapi-storage-script/main.ml

```diff
@@ -11,18 +11,293 @@
+          Lwt_list.iter_s (fun f -> f ()) acc
+      | Regular | Other | Unknown ->
+          let msg =
+            Printf.sprintf
+              {|Could not create directory "%s": already exists and it's not a directory|}
+              path
+          in
+          Lwt.fail (Failure msg)
+      | Missing ->
+          let parent = Filename.dirname path in
+          loop ((fun () -> Lwt_unix.mkdir path perm) :: acc) parent
+    in
+    loop [] path
+end
+
+module Signal = struct
+  type t = int
+
+  let to_string s = Fmt.(str "%a" Dump.signal s)
+end
+
+module Process : sig
+  module Output : sig
+    type exit_or_signal = Exit_non_zero of int | Signal of Signal.t
+
+    type t = {
+        exit_status: (unit, exit_or_signal) Result.t
+      ; stdout: string
+      ; stderr: string
+    }
+  end
+
+  val run : prog:string -> args:string list -> input:string -> Output.t Lwt.t
+  (** Runs a cli program, writes [input] into its stdin, then closing the fd,
+      and finally waits for the program to finish and returns the exit status,
+      its stdout and stderr. *)
+end = struct
+  module Output = struct
+    type exit_or_signal = Exit_non_zero of int | Signal of Signal.t
+
+    type t = {
+        exit_status: (unit, exit_or_signal) Result.t
+      ; stdout: string
+      ; stderr: string
+    }
+
+    let exit_or_signal_of_unix = function
+      | Unix.WEXITED 0 ->
+          Ok ()
+      | WEXITED n ->
+          Error (Exit_non_zero n)
+      | WSIGNALED n ->
+          Error (Signal n)
+      | WSTOPPED n ->
+          Error (Signal n)
+  end
+
+  type t = {proc: Lwt_process.process_full; closed: bool ref}
+
+  let create ~prog ~args =
+    let args = Array.of_list (prog :: args) in
+    let cmd = (prog, args) in
+
+    let proc = Lwt_process.open_process_full cmd in
+    {proc; closed= ref false}
+
+  (* Replicates Lwt_process' close, but allowing channels to be closed twice *)
+  let close p =
+    let close_chan_safe c =
+      Lwt.catch
+        (fun () -> Lwt_io.close c)
+        (function
+          | Lwt_io.Channel_closed _ -> Lwt.return_unit | exn -> raise exn
+          )
+    in
+    if !(p.closed) then
+      p.proc#status
+    else (
+      p.closed := true ;
+      Lwt.protected
+        (Lwt.join
+           (close_chan_safe p.proc#stdin
+           :: List.map close_chan_safe [p.proc#stdout; p.proc#stderr]
+           )
+        )
+      >>= fun () -> p.proc#status
+    )
+
+  let collect_output_and_wait p =
+    let read_all chan = Lwt_io.read chan in
+    Lwt.both (read_all p.proc#stdout) (read_all p.proc#stderr)
+    >>= fun (stdout, stderr) ->
+    close p >>= fun status ->
+    let exit_status = Output.exit_or_signal_of_unix status in
+    Lwt.return {Output.exit_status; stdout; stderr}
+
+  let run ~prog ~args ~input =
+    let p = create ~prog ~args in
+    Lwt.catch
+      (fun () ->
+        let w = p.proc#stdin in
+        Lwt_io.write w input >>= fun () -> Lwt_io.close w
```

Comment:
In theory this could deadlock: the process may write something to its output every time it consumes chunks from the input, and because these are pipes they have a limited capacity. Once full they block and the other side needs to consume them.
I think you need to write to stdin and read from stdout/stderr concurrently to prevent deadlocks.

Reply:
See how pmap is implemented: it starts both the getter and sender, and then it tries to wait for them (doesn't matter in what order because they'll keep running concurrently):[ ](https://github.com/ocsigen/lwt/blob/master/src/unix/lwt_process.cppo.ml#L501).

This is why I usually recommend to use `pmap` where possible, although if you need separate stdout and stderr then we don't have much choice but to reimplement it.

Reply:
I've reworked this, which makes for a simpler implementation, still needs to be tested.